// @progress/kendo-angular-grid/esm2022/pdf/pdf.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { IconsService, KENDO_ICONS } from '@progress/kendo-angular-icons';
import { KENDO_GRID_PDF_EXPORT } from '../directives';
import * as i0 from "@angular/core";
import * as i1 from "@progress/kendo-angular-icons";
import * as i2 from "./pdf.component";
import * as i3 from "./pdf-margin.component";
import * as i4 from "./pdf-command.directive";
import * as i5 from "./pdf-template.directive";
import * as i6 from "./pdf-command-tool.directive";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the Grid PDF component.
 *
 * @example
 *
 * ```ts
 * import { NgModule } from '@angular/core';
 * import { BrowserModule } from "@angular/platform-browser";
 * import { GridModule, PDFModule } from '@progress/kendo-angular-grid';
 * import { AppComponent } from './app.component';
 *
 * @NgModule({
 *     declarations: [AppComponent],
 *     imports:      [BrowserModule, GridModule, PDFModule],
 *     bootstrap:    [AppComponent]
 * })
 *
 * export class AppModule {}
 *
 * ```
 */
export class PDFModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: PDFModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: PDFModule, imports: [i1.IconComponent, i1.SVGIconComponent, i2.PDFComponent, i3.PDFMarginComponent, i4.PDFCommandDirective, i5.PDFTemplateDirective, i6.PDFCommandToolbarDirective], exports: [i2.PDFComponent, i3.PDFMarginComponent, i4.PDFCommandDirective, i5.PDFTemplateDirective, i6.PDFCommandToolbarDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: PDFModule, providers: [IconsService], imports: [KENDO_ICONS, i4.PDFCommandDirective] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: PDFModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...KENDO_ICONS, ...KENDO_GRID_PDF_EXPORT],
                    exports: [...KENDO_GRID_PDF_EXPORT],
                    providers: [IconsService]
                }]
        }] });
