// @progress/kendo-angular-buttons/esm2022/directives.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ButtonComponent } from './button/button.component';
import { ButtonGroupComponent } from './buttongroup/buttongroup.component';
import { ChipListComponent } from './chip/chip-list.component';
import { ChipComponent } from './chip/chip.component';
import { DropDownButtonComponent } from './dropdownbutton/dropdownbutton.component';
import { FloatingActionButtonComponent } from './floatingactionbutton/floatingactionbutton.component';
import { DialItemTemplateDirective } from './floatingactionbutton/templates/dial-item-template.directive';
import { FloatingActionButtonTemplateDirective } from './floatingactionbutton/templates/fab-template.directive';
import { SplitButtonCustomMessagesComponent } from './splitbutton/localization/custom-messages.component';
import { SplitButtonComponent } from './splitbutton/splitbutton.component';
import { ButtonItemTemplateDirective } from './listbutton/button-item-template.directive';
import { ToggleButtonTabStopDirective } from '@progress/kendo-angular-common';
/**
 * Utility array that contains all `Button` related components and directives
 */
export const KENDO_BUTTON = [
    ButtonComponent
];
/**
 * Utility array that contains all `ButtonGroup` related components and directives
 */
export const KENDO_BUTTONGROUP = [
    ButtonComponent,
    ButtonGroupComponent
];
/**
 * Utility array that contains all `DropDownButton` related components and directives
 */
export const KENDO_DROPDOWNBUTTON = [
    DropDownButtonComponent,
    ButtonItemTemplateDirective
];
/**
 * Utility array that contains all `Chip` related components and directives
 */
export const KENDO_CHIP = [
    ChipComponent
];
/**
 * Utility array that contains all `ChipList` related components and directives
 */
export const KENDO_CHIPLIST = [
    ChipComponent,
    ChipListComponent
];
/**
 * Utility array that contains all `FloatingActionButton` related components and directives
 */
export const KENDO_FLOATINGACTIONBUTTON = [
    FloatingActionButtonComponent,
    DialItemTemplateDirective,
    FloatingActionButtonTemplateDirective
];
/**
 * Utility array that contains all `SplitButton` related components and directives
 */
export const KENDO_SPLITBUTTON = [
    SplitButtonComponent,
    SplitButtonCustomMessagesComponent,
    ToggleButtonTabStopDirective
];
/**
 * Utility array that contains all `@progress/kendo-angular-buttons` related components and directives
 */
export const KENDO_BUTTONS = [
    ...KENDO_BUTTON,
    ...KENDO_BUTTONGROUP,
    ...KENDO_DROPDOWNBUTTON,
    ...KENDO_CHIP,
    ...KENDO_CHIPLIST,
    ...KENDO_FLOATINGACTIONBUTTON,
    ...KENDO_SPLITBUTTON
];
