// @progress/kendo-angular-grid/esm2022/excel/excel-command-tool.directive.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive } from '@angular/core';
import { ToolBarButtonComponent } from '@progress/kendo-angular-toolbar';
import { Subscription } from 'rxjs';
import { fileExcelIcon } from '@progress/kendo-svg-icons';
import { ExcelService } from './excel.service';
import * as i0 from "@angular/core";
import * as i1 from "./excel.service";
import * as i2 from "@progress/kendo-angular-toolbar";
/**
 * Represents the `export-to-Excel` toolbar tool of the Grid.
 * You can apply this directive to any `kendo-toolbar-button` element inside a
 * ToolbarComponent used in the Grid.
 *
 * When the user clicks a button associated with the directive, the
 * [excelExport]({% slug api_grid_gridcomponent %}#toc-excelexport) event
 * fires ([see example]({% slug excelexport_grid %})).
 *
 * @example
 * ```html-no-run
 * <kendo-grid>
 *      <kendo-toolbar>
 *         <kendo-toolbar-button text="Excel Export" kendoGridExcelTool></kendo-toolbar-button>
 *      </kendo-toolbar>
 *      <kendo-grid-excel fileName="Grid.xlsx">
 *      </kendo-grid-excel>
 * </kendo-grid>
 * ```
 */
export class ExcelCommandToolbarDirective {
    excelService;
    host;
    clickSub = new Subscription();
    constructor(excelService, host) {
        this.excelService = excelService;
        this.host = host;
    }
    ngOnInit() {
        this.clickSub = this.host.click.subscribe(e => this.onClick(e));
        this.host.className = 'k-grid-excel';
        this.host.svgIcon = fileExcelIcon;
        this.host.icon = 'file-excel';
    }
    ngOnDestroy() {
        this.clickSub.unsubscribe();
    }
    /**
     * @hidden
     */
    onClick(e) {
        e.preventDefault();
        this.excelService.exportClick.emit();
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ExcelCommandToolbarDirective, deps: [{ token: i1.ExcelService }, { token: i2.ToolBarButtonComponent }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: ExcelCommandToolbarDirective, isStandalone: true, selector: "[kendoGridExcelTool]", ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ExcelCommandToolbarDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[kendoGridExcelTool]',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ExcelService }, { type: i2.ToolBarButtonComponent }]; } });
