// @progress/kendo-angular-grid/esm2022/excel/excel.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { IconsService, KENDO_ICONS } from '@progress/kendo-angular-icons';
import { KENDO_EXCELEXPORT } from '@progress/kendo-angular-excel-export';
import { KENDO_GRID_EXCEL_EXPORT } from '../directives';
import * as i0 from "@angular/core";
import * as i1 from "@progress/kendo-angular-icons";
import * as i2 from "./excel.component";
import * as i3 from "./excel-command.directive";
import * as i4 from "./excel-command-tool.directive";
import * as i5 from "@progress/kendo-angular-excel-export";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the Excel component of the Grid.
 *
 * @example
 *
 * ```ts
 * import { NgModule } from '@angular/core';
 * import { BrowserModule } from "@angular/platform-browser";
 * import { GridModule, ExcelModule } from '@progress/kendo-angular-grid';
 * import { AppComponent } from './app.component';
 *
 * _@NgModule({
 *     declarations: [AppComponent],
 *     imports:      [BrowserModule, GridModule, ExcelModule],
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 *
 * ```
 */
export class ExcelModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ExcelModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: ExcelModule, imports: [i1.IconComponent, i1.SVGIconComponent, i2.ExcelComponent, i3.ExcelCommandDirective, i4.ExcelCommandToolbarDirective, i5.ExcelExportComponent, i5.ColumnComponent, i5.ColumnGroupComponent, i5.FooterTemplateDirective, i5.GroupFooterTemplateDirective, i5.GroupHeaderColumnTemplateDirective, i5.GroupHeaderTemplateDirective], exports: [i2.ExcelComponent, i3.ExcelCommandDirective, i4.ExcelCommandToolbarDirective, i5.ExcelExportComponent, i5.ColumnComponent, i5.ColumnGroupComponent, i5.FooterTemplateDirective, i5.GroupFooterTemplateDirective, i5.GroupHeaderColumnTemplateDirective, i5.GroupHeaderTemplateDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ExcelModule, providers: [IconsService], imports: [KENDO_ICONS, i3.ExcelCommandDirective, i5.ExcelExportComponent, i5.ColumnComponent, i5.ColumnGroupComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ExcelModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...KENDO_ICONS, ...KENDO_GRID_EXCEL_EXPORT, ...KENDO_EXCELEXPORT],
                    exports: [...KENDO_GRID_EXCEL_EXPORT, ...KENDO_EXCELEXPORT],
                    providers: [IconsService]
                }]
        }] });
