// @progress/kendo-angular-dateinputs/esm2022/dateinput/dateinput.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_DATEINPUT } from '../directives';
import { IconsService } from '@progress/kendo-angular-icons';
import * as i0 from "@angular/core";
import * as i1 from "./localization/dateinput-custom-messages.component";
import * as i2 from "./dateinput.component";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the DateInput component.
 */
export class DateInputModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: DateInputModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: DateInputModule, imports: [i1.DateInputCustomMessagesComponent, i2.DateInputComponent], exports: [i1.DateInputCustomMessagesComponent, i2.DateInputComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: DateInputModule, providers: [
            IconsService
        ], imports: [i2.DateInputComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: DateInputModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_DATEINPUT],
                    imports: [...KENDO_DATEINPUT],
                    providers: [
                        IconsService
                    ]
                }]
        }] });
