// @progress/kendo-angular-navigation/esm2022/actionsheet.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { IconsService } from '@progress/kendo-angular-icons';
import { KENDO_ACTIONSHEET } from './directives';
import * as i0 from "@angular/core";
import * as i1 from "./actionsheet/actionsheet.component";
import * as i2 from "./actionsheet/templates/header-template.directive";
import * as i3 from "./actionsheet/templates/item-template.directive";
import * as i4 from "./actionsheet/templates/content-template.directive";
import * as i5 from "./actionsheet/templates/footer-template.directive";
import * as i6 from "./actionsheet/templates/actionsheet-template";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the ActionSheet component.
 *
 * @example
 *
 * ```ts-no-run
 * // Import the ActionSheet module
 * import { ActionSheetModule } from '@progress/kendo-angular-navigation';
 *
 * // The browser platform with a compiler
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 * import { BrowserModule } from '@angular/platform-browser';
 *
 * import { NgModule } from '@angular/core';
 *
 * // Import the app component
 * import { AppComponent } from './app.component';
 *
 * // Define the app module
 * _@NgModule({
 *     declarations: [AppComponent], // declare app component
 *     imports:      [BrowserModule, ActionSheetModule], // import ActionSheet module
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 *
 * // Compile and launch the module
 * platformBrowserDynamic().bootstrapModule(AppModule);
 *
 * ```
 */
export class ActionSheetModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ActionSheetModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: ActionSheetModule, imports: [i1.ActionSheetComponent, i2.ActionSheetHeaderTemplateDirective, i3.ActionSheetItemTemplateDirective, i4.ActionSheetContentTemplateDirective, i5.ActionSheetFooterTemplateDirective, i6.ActionSheetTemplateDirective], exports: [i1.ActionSheetComponent, i2.ActionSheetHeaderTemplateDirective, i3.ActionSheetItemTemplateDirective, i4.ActionSheetContentTemplateDirective, i5.ActionSheetFooterTemplateDirective, i6.ActionSheetTemplateDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ActionSheetModule, providers: [IconsService], imports: [i1.ActionSheetComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ActionSheetModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_ACTIONSHEET],
                    imports: [...KENDO_ACTIONSHEET],
                    providers: [IconsService]
                }]
        }] });
