// @progress/kendo-angular-excel-export/esm2022/excel-export.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_EXCELEXPORT } from './directives';
import * as i0 from "@angular/core";
import * as i1 from "./excel-export.component";
import * as i2 from "./columns/column.component";
import * as i3 from "./columns/column-group.component";
import * as i4 from "./columns/footer-template.directive";
import * as i5 from "./columns/group-footer-template.directive";
import * as i6 from "./columns/group-header-column-template.directive";
import * as i7 from "./columns/group-header-template.directive";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the Excel Export component.
 *
 * @example
 *
 * ```ts-no-run
 * // Import the ExcelExportModule module
 * import { ExcelExportModule } from '@progress/kendo-angular-excel-export';
 *
 * // The browser platform with a compiler
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 *
 * import { NgModule } from '@angular/core';
 *
 * // Import the app component
 * import { AppComponent } from './app.component';
 *
 * // Define the app module
 * _@NgModule({
 *     declarations: [AppComponent], // declare app component
 *     imports:      [BrowserModule, ExcelExportModule], // import ExcelExportModule module
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 *
 * // Compile and launch the module
 * platformBrowserDynamic().bootstrapModule(AppModule);
 *
 * ```
 */
export class ExcelExportModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ExcelExportModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: ExcelExportModule, imports: [i1.ExcelExportComponent, i2.ColumnComponent, i3.ColumnGroupComponent, i4.FooterTemplateDirective, i5.GroupFooterTemplateDirective, i6.GroupHeaderColumnTemplateDirective, i7.GroupHeaderTemplateDirective], exports: [i1.ExcelExportComponent, i2.ColumnComponent, i3.ColumnGroupComponent, i4.FooterTemplateDirective, i5.GroupFooterTemplateDirective, i6.GroupHeaderColumnTemplateDirective, i7.GroupHeaderTemplateDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ExcelExportModule });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ExcelExportModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...KENDO_EXCELEXPORT],
                    exports: [...KENDO_EXCELEXPORT]
                }]
        }] });
