// @progress/kendo-angular-dateinputs/esm2022/datepicker/datepicker.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_DATEPICKER } from '../directives';
import { IconsService } from '@progress/kendo-angular-icons';
import { KENDO_TOGGLEBUTTONTABSTOP, ResizeBatchService } from '@progress/kendo-angular-common';
import { PopupService } from '@progress/kendo-angular-popup';
import { CenturyViewService } from '../calendar/services/century-view.service';
import { DecadeViewService } from '../calendar/services/decade-view.service';
import { CalendarDOMService } from '../calendar/services/dom.service';
import { MonthViewService } from '../calendar/services/month-view.service';
import { NavigationService } from '../calendar/services/navigation.service';
import { YearViewService } from '../calendar/services/year-view.service';
import * as i0 from "@angular/core";
import * as i1 from "./localization/datepicker-custom-messages.component";
import * as i2 from "./datepicker.component";
import * as i3 from "../calendar/templates/cell-template.directive";
import * as i4 from "../calendar/templates/century-cell-template.directive";
import * as i5 from "../calendar/templates/decade-cell-template.directive";
import * as i6 from "../calendar/templates/footer-template.directive";
import * as i7 from "../calendar/templates/header-template.directive";
import * as i8 from "../calendar/templates/header-title-template.directive";
import * as i9 from "../calendar/templates/month-cell-template.directive";
import * as i10 from "../calendar/templates/navigation-item-template.directive";
import * as i11 from "../calendar/templates/weeknumber-cell-template.directive";
import * as i12 from "../calendar/templates/year-cell-template.directive";
import * as i13 from "@progress/kendo-angular-common";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the DatePicker component.
 */
export class DatePickerModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: DatePickerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: DatePickerModule, imports: [i1.DatePickerCustomMessagesComponent, i2.DatePickerComponent, i3.CellTemplateDirective, i4.CenturyCellTemplateDirective, i5.DecadeCellTemplateDirective, i6.FooterTemplateDirective, i7.HeaderTemplateDirective, i8.HeaderTitleTemplateDirective, i9.MonthCellTemplateDirective, i10.NavigationItemTemplateDirective, i11.WeekNumberCellTemplateDirective, i12.YearCellTemplateDirective, i10.NavigationItemTemplateDirective, i13.ToggleButtonTabStopDirective, i13.ToggleButtonTabStopDirective], exports: [i1.DatePickerCustomMessagesComponent, i2.DatePickerComponent, i3.CellTemplateDirective, i4.CenturyCellTemplateDirective, i5.DecadeCellTemplateDirective, i6.FooterTemplateDirective, i7.HeaderTemplateDirective, i8.HeaderTitleTemplateDirective, i9.MonthCellTemplateDirective, i10.NavigationItemTemplateDirective, i11.WeekNumberCellTemplateDirective, i12.YearCellTemplateDirective, i10.NavigationItemTemplateDirective, i13.ToggleButtonTabStopDirective, i13.ToggleButtonTabStopDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: DatePickerModule, providers: [
            // DateInputModule providers
            IconsService,
            // CalendarModule providers
            ResizeBatchService,
            CalendarDOMService,
            CenturyViewService,
            DecadeViewService,
            MonthViewService,
            YearViewService,
            PopupService,
            NavigationService
        ], imports: [i2.DatePickerComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: DatePickerModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_DATEPICKER, ...KENDO_TOGGLEBUTTONTABSTOP],
                    imports: [...KENDO_DATEPICKER, ...KENDO_TOGGLEBUTTONTABSTOP],
                    providers: [
                        // DateInputModule providers
                        IconsService,
                        // CalendarModule providers
                        ResizeBatchService,
                        CalendarDOMService,
                        CenturyViewService,
                        DecadeViewService,
                        MonthViewService,
                        YearViewService,
                        PopupService,
                        NavigationService
                    ]
                }]
        }] });
