// @progress/kendo-angular-pdf-export/esm2022/pdf-export-margin.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Component, Input } from '@angular/core';
import * as i0 from "@angular/core";
const FIELDS = ['bottom', 'left', 'right', 'top'];
/**
 * Represents the Kendo UI PDFMargin component for Angular.
 */
export class PDFExportMarginComponent {
    /**
     * The bottom margin. The supported units are `"mm"`, `"cm"`, `"in"`, and `"pt"` (default).
     * Numbers are considered to be points (`"pt"`).
     */
    left;
    /**
     * The top margin. The supported units are `"mm"`, `"cm"`, `"in"`, and `"pt"` (default).
     * Numbers are considered to be points (`"pt"`).
     */
    top;
    /**
     * The right margin. The supported units are `"mm"`, `"cm"`, `"in"`, and `"pt"` (default).
     * Numbers are considered to be points (`"pt"`).
     */
    right;
    /**
     * The bottom margin. The supported units are `"mm"`, `"cm"`, `"in"`, and `"pt"` (default).
     * Numbers are considered to be points (`"pt"`).
     */
    bottom;
    /**
     * @hidden
     */
    get options() {
        const options = {};
        for (let idx = 0; idx < FIELDS.length; idx++) {
            const field = FIELDS[idx];
            const value = this[field];
            if (typeof value !== 'undefined') {
                options[field] = value;
            }
        }
        return options;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: PDFExportMarginComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: PDFExportMarginComponent, isStandalone: true, selector: "kendo-pdf-export-margin", inputs: { left: "left", top: "top", right: "right", bottom: "bottom" }, ngImport: i0, template: ``, isInline: true });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: PDFExportMarginComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'kendo-pdf-export-margin',
                    template: ``,
                    standalone: true
                }]
        }], propDecorators: { left: [{
                type: Input
            }], top: [{
                type: Input
            }], right: [{
                type: Input
            }], bottom: [{
                type: Input
            }] } });
