// @progress/kendo-angular-inputs/esm2022/directives.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { PrefixTemplateDirective, SuffixTemplateDirective, SeparatorComponent } from "@progress/kendo-angular-common";
import { CheckBoxComponent } from "./checkbox/checkbox.component";
import { CheckBoxDirective } from "./checkbox/checkbox.directive";
import { ColorGradientComponent } from "./colorpicker/color-gradient.component";
import { ColorPaletteComponent } from "./colorpicker/color-palette.component";
import { ColorPickerComponent } from "./colorpicker/colorpicker.component";
import { FlatColorPickerComponent } from "./colorpicker/flatcolorpicker.component";
import { ColorPickerCustomMessagesComponent } from "./colorpicker/localization/custom-messages.component";
import { ErrorComponent } from "./formfield/error.component";
import { FormFieldComponent } from "./formfield/formfield.component";
import { HintComponent } from "./formfield/hint.component";
import { MaskedTextBoxComponent } from "./maskedtextbox/maskedtextbox.component";
import { NumericTextBoxCustomMessagesComponent } from "./numerictextbox/localization/custom-messages.component";
import { NumericTextBoxComponent } from "./numerictextbox/numerictextbox.component";
import { RadioButtonComponent } from "./radiobutton/radiobutton.component";
import { RadioButtonDirective } from "./radiobutton/radiobutton.directive";
import { RangeSliderCustomMessagesComponent } from "./rangeslider/localization/custom-messages.component";
import { RangeSliderComponent } from "./rangeslider/rangeslider.component";
import { RatingHoveredItemTemplateDirective } from "./rating/directives/rating-hovered-item.directive";
import { RatingItemTemplateDirective } from "./rating/directives/rating-item.directive";
import { RatingSelectedItemTemplateDirective } from "./rating/directives/rating-selected-item.directive";
import { RatingComponent } from "./rating/rating.component";
import { InputSeparatorComponent } from "./shared/input-separator.component";
import { TextAreaDirective } from "./shared/textarea.directive";
import { SignatureCustomMessagesComponent } from "./signature/localization/custom-messages.component";
import { SignatureComponent } from "./signature/signature.component";
import { SliderCustomMessagesComponent } from "./slider/localization/custom-messages.component";
import { SliderComponent } from "./slider/slider.component";
import { LabelTemplateDirective } from "./sliders-common/label-template.directive";
import { SwitchCustomMessagesComponent } from "./switch/localization/custom-messages.component";
import { SwitchComponent } from "./switch/switch.component";
import { TextAreaPrefixComponent } from "./textarea/textarea-prefix.component";
import { TextAreaSuffixComponent } from "./textarea/textarea-suffix.component";
import { TextAreaComponent } from "./textarea/textarea.component";
import { TextBoxCustomMessagesComponent } from "./textbox/localization/custom-messages.component";
import { TextBoxPrefixTemplateDirective } from "./textbox/textbox-prefix.directive";
import { TextBoxSuffixTemplateDirective } from "./textbox/textbox-suffix.directive";
import { TextBoxComponent } from "./textbox/textbox.component";
import { TextBoxDirective } from "./textbox/textbox.directive";
/**
 * Utility array that contains all `TextBox` related components and directives
 */
export const KENDO_TEXTBOX = [
    TextBoxDirective,
    TextBoxComponent,
    InputSeparatorComponent,
    TextBoxSuffixTemplateDirective,
    TextBoxPrefixTemplateDirective,
    TextBoxCustomMessagesComponent,
    PrefixTemplateDirective,
    SuffixTemplateDirective,
    SeparatorComponent
];
/**
 * Utility array that contains all `NumericTextBox` related components and directives
 */
export const KENDO_NUMERICTEXTBOX = [
    NumericTextBoxComponent,
    NumericTextBoxCustomMessagesComponent,
    PrefixTemplateDirective,
    SuffixTemplateDirective,
    SeparatorComponent
];
/**
 * Utility array that contains all `MaskedTextBox` related components and directives
 */
export const KENDO_MASKEDTEXTBOX = [
    MaskedTextBoxComponent,
    PrefixTemplateDirective,
    SuffixTemplateDirective,
    SeparatorComponent
];
/**
 * Utility array that contains all `TextArea` related components and directives
 */
export const KENDO_TEXTAREA = [
    TextAreaComponent,
    TextAreaDirective,
    TextAreaPrefixComponent,
    TextAreaSuffixComponent,
    SeparatorComponent
];
/**
 * Utility array that contains all `CheckBox` related components and directives
 */
export const KENDO_CHECKBOX = [
    CheckBoxComponent,
    CheckBoxDirective
];
/**
 * Utility array that contains all `RadioButton` related components and directives
 */
export const KENDO_RADIOBUTTON = [
    RadioButtonComponent,
    RadioButtonDirective
];
/**
 * Utility array that contains all `Switch` related components and directives
 */
export const KENDO_SWITCH = [
    SwitchComponent,
    SwitchCustomMessagesComponent
];
/**
 * Utility array that contains all `FormField` related components and directives
 */
export const KENDO_FORMFIELD = [
    FormFieldComponent,
    HintComponent,
    ErrorComponent
];
/**
 * Utility array that contains all `Slider` related components and directives
 */
export const KENDO_SLIDER = [
    SliderComponent,
    SliderCustomMessagesComponent,
    LabelTemplateDirective,
];
/**
 * Utility array that contains all `RangeSlider` related components and directives
 */
export const KENDO_RANGESLIDER = [
    RangeSliderComponent,
    RangeSliderCustomMessagesComponent,
    LabelTemplateDirective
];
/**
 * Utility array that contains all `Rating` related components and directives
 */
export const KENDO_RATING = [
    RatingComponent,
    RatingItemTemplateDirective,
    RatingHoveredItemTemplateDirective,
    RatingSelectedItemTemplateDirective
];
/**
 * Utility array that contains all `Signature` related components and directives
 */
export const KENDO_SIGNATURE = [
    SignatureComponent,
    SignatureCustomMessagesComponent
];
/**
 * Utility array that contains all `ColorPicker` related components and directives
 */
export const KENDO_COLORPICKER = [
    ColorPickerComponent,
    ColorPickerCustomMessagesComponent
];
/**
 * Utility array that contains all `FlatColorPicker` related components and directives
 */
export const KENDO_FLATCOLORPICKER = [
    FlatColorPickerComponent,
    ColorPickerCustomMessagesComponent
];
/**
 * Utility array that contains all `ColorPallete` related components and directives
 */
export const KENDO_COLORPALETTE = [
    ColorPaletteComponent,
    ColorPickerCustomMessagesComponent
];
/**
 * Utility array that contains all `ColorGradient` related components and directives
 */
export const KENDO_COLORGRADIENT = [
    ColorGradientComponent,
    ColorPickerCustomMessagesComponent
];
/**
 * Utility array that contains all `@progress/kendo-angular-inputs` related components and directives
 */
export const KENDO_INPUTS = [
    ...KENDO_TEXTBOX,
    ...KENDO_NUMERICTEXTBOX,
    ...KENDO_MASKEDTEXTBOX,
    ...KENDO_TEXTAREA,
    ...KENDO_CHECKBOX,
    ...KENDO_RADIOBUTTON,
    ...KENDO_SWITCH,
    ...KENDO_FORMFIELD,
    ...KENDO_SLIDER,
    ...KENDO_RANGESLIDER,
    ...KENDO_RATING,
    ...KENDO_SIGNATURE,
    ...KENDO_COLORPICKER,
    ...KENDO_FLATCOLORPICKER,
    ...KENDO_COLORGRADIENT,
    ...KENDO_COLORPALETTE,
];
