// @progress/kendo-angular-inputs/esm2022/maskedtextbox.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_MASKEDTEXTBOX } from './directives';
import { KENDO_ADORNMENTS } from '@progress/kendo-angular-common';
import * as i0 from "@angular/core";
import * as i1 from "./maskedtextbox/maskedtextbox.component";
import * as i2 from "@progress/kendo-angular-common";
//IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the MaskedTextBox component.
 *
 * @example
 *
 * ```ts-no-run
 * // Import the MaskedTextBox module
 * import { MaskedTextBoxModule } from '@progress/kendo-angular-inputs';
 *
 * // The browser platform with a compiler
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 *
 * import { NgModule } from '@angular/core';
 *
 * // Import the app component
 * import { AppComponent } from './app.component';
 *
 * // Define the app module
 * _@NgModule({
 *     declarations: [AppComponent], // declare app component
 *     imports:      [BrowserModule, MaskedTextBoxModule], // import MaskedTextBox module
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 *
 * // Compile and launch the module
 * platformBrowserDynamic().bootstrapModule(AppModule);
 *
 * ```
 */
export class MaskedTextBoxModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: MaskedTextBoxModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: MaskedTextBoxModule, imports: [i1.MaskedTextBoxComponent, i2.PrefixTemplateDirective, i2.SuffixTemplateDirective, i2.SeparatorComponent, i2.PrefixTemplateDirective, i2.SuffixTemplateDirective, i2.SeparatorComponent], exports: [i1.MaskedTextBoxComponent, i2.PrefixTemplateDirective, i2.SuffixTemplateDirective, i2.SeparatorComponent, i2.PrefixTemplateDirective, i2.SuffixTemplateDirective, i2.SeparatorComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: MaskedTextBoxModule, imports: [i2.SeparatorComponent, i2.SeparatorComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: MaskedTextBoxModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...KENDO_MASKEDTEXTBOX, ...KENDO_ADORNMENTS],
                    exports: [...KENDO_MASKEDTEXTBOX, ...KENDO_ADORNMENTS]
                }]
        }] });
