// @progress/kendo-angular-inputs/esm2022/textarea.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from "@angular/core";
import { KENDO_TEXTAREA } from "./directives";
import { KENDO_ADORNMENTS } from "@progress/kendo-angular-common";
import * as i0 from "@angular/core";
import * as i1 from "./textarea/textarea.component";
import * as i2 from "./shared/textarea.directive";
import * as i3 from "./textarea/textarea-prefix.component";
import * as i4 from "./textarea/textarea-suffix.component";
import * as i5 from "@progress/kendo-angular-common";
//IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the TextArea component.
 *
 * @example
 *
 * ```ts-no-run
 * // Import the TextArea module
 * import { TextAreaModule } from '@progress/kendo-angular-inputs';
 *
 * // The browser platform with a compiler
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 *
 * import { NgModule } from '@angular/core';
 *
 * // Import the app component
 * import { AppComponent } from './app.component';
 *
 * // Define the app module
 * _@NgModule({
 *     declarations: [AppComponent], // declare app component
 *     imports:      [BrowserModule, TextAreaModule], // import TextArea module
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 *
 * // Compile and launch the module
 * platformBrowserDynamic().bootstrapModule(AppModule);
 *
 * ```
 */
export class TextAreaModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TextAreaModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TextAreaModule, imports: [i1.TextAreaComponent, i2.TextAreaDirective, i3.TextAreaPrefixComponent, i4.TextAreaSuffixComponent, i5.SeparatorComponent, i5.PrefixTemplateDirective, i5.SuffixTemplateDirective, i5.SeparatorComponent], exports: [i1.TextAreaComponent, i2.TextAreaDirective, i3.TextAreaPrefixComponent, i4.TextAreaSuffixComponent, i5.SeparatorComponent, i5.PrefixTemplateDirective, i5.SuffixTemplateDirective, i5.SeparatorComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TextAreaModule, imports: [i5.SeparatorComponent, i5.SeparatorComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TextAreaModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...KENDO_TEXTAREA, ...KENDO_ADORNMENTS],
                    exports: [...KENDO_TEXTAREA, ...KENDO_ADORNMENTS]
                }]
        }] });
