// @progress/kendo-angular-dateinputs/esm2022/daterange/date-range.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_DATERANGE } from '../directives';
import { PopupService } from '@progress/kendo-angular-popup';
import { CenturyViewService } from '../calendar/services/century-view.service';
import { DecadeViewService } from '../calendar/services/decade-view.service';
import { MonthViewService } from '../calendar/services/month-view.service';
import { NavigationService } from '../calendar/services/navigation.service';
import { YearViewService } from '../calendar/services/year-view.service';
import { IconsService } from '@progress/kendo-angular-icons';
import { ResizeBatchService } from '@progress/kendo-angular-common';
import * as i0 from "@angular/core";
import * as i1 from "./localization/daterange-popup-custom-messages.component";
import * as i2 from "../dateinput/localization/dateinput-custom-messages.component";
import * as i3 from "../dateinput/dateinput.component";
import * as i4 from "./date-range-end-input.directive";
import * as i5 from "./date-range-popup-template.directive";
import * as i6 from "./date-range.component";
import * as i7 from "./date-range-popup.component";
import * as i8 from "./date-range-selection.directive";
import * as i9 from "./date-range-start-input.directive";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * The exported package module.
 *
 * The package exports:
 * - `DateRangeComponent`&mdash;The DateRange component class.
 * - `DateRangePopupComponent`&mdash;The DateRangePopup component class.
 * - `DateRangeSelectionDirective`&mdash;The MultiviewCalendar date range selection directive.
 * - `DateRangeEndInputDirective`&mdash;The end DateInput date range selection directive.
 * - `DateRangeStartInputDirective`&mdash;The start DateInput date range selection directive.
 * - `DateRangePopupTemplateDirective`&mdash;The DateRangePopup content template directive.
 *
 * @example
 *
 * ```ts-no-run
 * // Import the DateRange module
 * import { DateRangeModule } from '@progress/kendo-angular-dateinputs';
 *
 * // The browser platform with a compiler
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 *
 * import { NgModule } from '@angular/core';
 *
 * // Import the app component
 * import { AppComponent } from './app.component';
 *
 * // Define the app module
 * _@NgModule({
 *     declarations: [AppComponent], // declare app component
 *     imports:      [BrowserModule, DateRangeModule], // import DateRange module
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 *
 * // Compile and launch the module
 * platformBrowserDynamic().bootstrapModule(AppModule);
 *
 * ```
 */
export class DateRangeModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: DateRangeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: DateRangeModule, imports: [i1.DateRangePopupCustomMessagesComponent, i2.DateInputCustomMessagesComponent, i3.DateInputComponent, i4.DateRangeEndInputDirective, i5.DateRangePopupTemplateDirective, i6.DateRangeComponent, i7.DateRangePopupComponent, i8.DateRangeSelectionDirective, i9.DateRangeStartInputDirective], exports: [i1.DateRangePopupCustomMessagesComponent, i2.DateInputCustomMessagesComponent, i3.DateInputComponent, i4.DateRangeEndInputDirective, i5.DateRangePopupTemplateDirective, i6.DateRangeComponent, i7.DateRangePopupComponent, i8.DateRangeSelectionDirective, i9.DateRangeStartInputDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: DateRangeModule, providers: [
            // MultiViewCalendarModule providers
            PopupService,
            NavigationService,
            CenturyViewService,
            DecadeViewService,
            MonthViewService,
            YearViewService,
            // DateInputModule providers
            IconsService,
            // AdaptiveModule providers
            ResizeBatchService
        ], imports: [i3.DateInputComponent, i6.DateRangeComponent, i7.DateRangePopupComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: DateRangeModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_DATERANGE],
                    imports: [...KENDO_DATERANGE],
                    providers: [
                        // MultiViewCalendarModule providers
                        PopupService,
                        NavigationService,
                        CenturyViewService,
                        DecadeViewService,
                        MonthViewService,
                        YearViewService,
                        // DateInputModule providers
                        IconsService,
                        // AdaptiveModule providers
                        ResizeBatchService
                    ]
                }]
        }] });
