// @progress/kendo-angular-dateinputs/esm2022/datetimepicker/datetimepicker.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_DATETIMEPICKER } from '../directives';
import { IconsService } from '@progress/kendo-angular-icons';
import { KENDO_TOGGLEBUTTONTABSTOP, ResizeBatchService } from '@progress/kendo-angular-common';
import { PopupService } from '@progress/kendo-angular-popup';
import { CenturyViewService } from '../calendar/services/century-view.service';
import { DecadeViewService } from '../calendar/services/decade-view.service';
import { CalendarDOMService } from '../calendar/services/dom.service';
import { MonthViewService } from '../calendar/services/month-view.service';
import { NavigationService } from '../calendar/services/navigation.service';
import { YearViewService } from '../calendar/services/year-view.service';
import { DayPeriodService } from '../timepicker/services/dayperiod.service';
import { TimePickerDOMService } from '../timepicker/services/dom.service';
import { HoursService } from '../timepicker/services/hours.service';
import { MillisecondsService } from '../timepicker/services/milliseconds.service';
import { MinutesService } from '../timepicker/services/minutes.service';
import { SecondsService } from '../timepicker/services/seconds.service';
import * as i0 from "@angular/core";
import * as i1 from "./localization/datetimepicker-custom-messages.component";
import * as i2 from "./datetimepicker.component";
import * as i3 from "../calendar/templates/cell-template.directive";
import * as i4 from "../calendar/templates/century-cell-template.directive";
import * as i5 from "../calendar/templates/decade-cell-template.directive";
import * as i6 from "../calendar/templates/footer-template.directive";
import * as i7 from "../calendar/templates/header-template.directive";
import * as i8 from "../calendar/templates/header-title-template.directive";
import * as i9 from "../calendar/templates/month-cell-template.directive";
import * as i10 from "../calendar/templates/navigation-item-template.directive";
import * as i11 from "../calendar/templates/weeknumber-cell-template.directive";
import * as i12 from "../calendar/templates/year-cell-template.directive";
import * as i13 from "@progress/kendo-angular-common";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the DateTimePicker component.
 */
export class DateTimePickerModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: DateTimePickerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: DateTimePickerModule, imports: [i1.DateTimePickerCustomMessagesComponent, i2.DateTimePickerComponent, i3.CellTemplateDirective, i4.CenturyCellTemplateDirective, i5.DecadeCellTemplateDirective, i6.FooterTemplateDirective, i7.HeaderTemplateDirective, i8.HeaderTitleTemplateDirective, i9.MonthCellTemplateDirective, i10.NavigationItemTemplateDirective, i11.WeekNumberCellTemplateDirective, i12.YearCellTemplateDirective, i13.ToggleButtonTabStopDirective, i13.ToggleButtonTabStopDirective], exports: [i1.DateTimePickerCustomMessagesComponent, i2.DateTimePickerComponent, i3.CellTemplateDirective, i4.CenturyCellTemplateDirective, i5.DecadeCellTemplateDirective, i6.FooterTemplateDirective, i7.HeaderTemplateDirective, i8.HeaderTitleTemplateDirective, i9.MonthCellTemplateDirective, i10.NavigationItemTemplateDirective, i11.WeekNumberCellTemplateDirective, i12.YearCellTemplateDirective, i13.ToggleButtonTabStopDirective, i13.ToggleButtonTabStopDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: DateTimePickerModule, providers: [
            // CalendarModule providers
            IconsService,
            ResizeBatchService,
            CalendarDOMService,
            CenturyViewService,
            DecadeViewService,
            MonthViewService,
            YearViewService,
            PopupService,
            NavigationService,
            // TimePickerModule providers
            TimePickerDOMService,
            HoursService,
            MinutesService,
            SecondsService,
            MillisecondsService,
            DayPeriodService,
        ], imports: [i2.DateTimePickerComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: DateTimePickerModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_DATETIMEPICKER, ...KENDO_TOGGLEBUTTONTABSTOP],
                    imports: [...KENDO_DATETIMEPICKER, ...KENDO_TOGGLEBUTTONTABSTOP],
                    providers: [
                        // CalendarModule providers
                        IconsService,
                        ResizeBatchService,
                        CalendarDOMService,
                        CenturyViewService,
                        DecadeViewService,
                        MonthViewService,
                        YearViewService,
                        PopupService,
                        NavigationService,
                        // TimePickerModule providers
                        TimePickerDOMService,
                        HoursService,
                        MinutesService,
                        SecondsService,
                        MillisecondsService,
                        DayPeriodService,
                    ]
                }]
        }] });
