// @progress/kendo-angular-dateinputs/esm2022/timepicker/timepicker.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_TIMEPICKER } from '../directives';
import { DayPeriodService } from './services/dayperiod.service';
import { TimePickerDOMService } from './services/dom.service';
import { HoursService } from './services/hours.service';
import { MillisecondsService } from './services/milliseconds.service';
import { MinutesService } from './services/minutes.service';
import { SecondsService } from './services/seconds.service';
import { IconsService } from '@progress/kendo-angular-icons';
import { PopupService } from '@progress/kendo-angular-popup';
import { KENDO_TOGGLEBUTTONTABSTOP, ResizeBatchService } from '@progress/kendo-angular-common';
import * as i0 from "@angular/core";
import * as i1 from "./localization/timepicker-custom-messages.component";
import * as i2 from "./timepicker.component";
import * as i3 from "@progress/kendo-angular-common";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the TimePicker component.
 */
export class TimePickerModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TimePickerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TimePickerModule, imports: [i1.TimePickerCustomMessagesComponent, i2.TimePickerComponent, i3.ToggleButtonTabStopDirective, i3.ToggleButtonTabStopDirective], exports: [i1.TimePickerCustomMessagesComponent, i2.TimePickerComponent, i3.ToggleButtonTabStopDirective, i3.ToggleButtonTabStopDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TimePickerModule, providers: [
            IconsService,
            PopupService,
            ResizeBatchService,
            TimePickerDOMService,
            HoursService,
            MinutesService,
            SecondsService,
            MillisecondsService,
            DayPeriodService,
        ], imports: [i2.TimePickerComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TimePickerModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_TIMEPICKER, ...KENDO_TOGGLEBUTTONTABSTOP],
                    imports: [...KENDO_TIMEPICKER, ...KENDO_TOGGLEBUTTONTABSTOP],
                    providers: [
                        IconsService,
                        PopupService,
                        ResizeBatchService,
                        TimePickerDOMService,
                        HoursService,
                        MinutesService,
                        SecondsService,
                        MillisecondsService,
                        DayPeriodService,
                    ]
                }]
        }] });
