// @progress/kendo-angular-dateinputs/esm2022/directives.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { CalendarComponent } from "./calendar/calendar.component";
import { MultiViewCalendarComponent } from "./calendar/multiview-calendar.component";
import { DateInputComponent } from "./dateinput/dateinput.component";
import { DatePickerComponent } from "./datepicker/datepicker.component";
import { DateRangePopupComponent } from "./daterange/date-range-popup.component";
import { DateRangeComponent } from "./daterange/date-range.component";
import { DateTimePickerComponent } from "./datetimepicker/datetimepicker.component";
import { TimePickerComponent } from "./timepicker/timepicker.component";
import { CellTemplateDirective } from "./calendar/templates/cell-template.directive";
import { CenturyCellTemplateDirective } from "./calendar/templates/century-cell-template.directive";
import { DecadeCellTemplateDirective } from "./calendar/templates/decade-cell-template.directive";
import { FooterTemplateDirective } from "./calendar/templates/footer-template.directive";
import { HeaderTemplateDirective } from "./calendar/templates/header-template.directive";
import { HeaderTitleTemplateDirective } from "./calendar/templates/header-title-template.directive";
import { MonthCellTemplateDirective } from "./calendar/templates/month-cell-template.directive";
import { NavigationItemTemplateDirective } from "./calendar/templates/navigation-item-template.directive";
import { WeekNumberCellTemplateDirective } from "./calendar/templates/weeknumber-cell-template.directive";
import { YearCellTemplateDirective } from "./calendar/templates/year-cell-template.directive";
import { DateRangeEndInputDirective } from "./daterange/date-range-end-input.directive";
import { DateRangePopupTemplateDirective } from "./daterange/date-range-popup-template.directive";
import { DateRangeSelectionDirective } from "./daterange/date-range-selection.directive";
import { DateRangeStartInputDirective } from "./daterange/date-range-start-input.directive";
import { CalendarCustomMessagesComponent } from "./calendar/localization/calendar-custom-messages.component";
import { MultiViewCalendarCustomMessagesComponent } from "./calendar/localization/multiview-calendar-custom-messages.component";
import { DateInputCustomMessagesComponent } from "./dateinput/localization/dateinput-custom-messages.component";
import { DatePickerCustomMessagesComponent } from "./datepicker/localization/datepicker-custom-messages.component";
import { DateRangePopupCustomMessagesComponent } from "./daterange/localization/daterange-popup-custom-messages.component";
import { DateTimePickerCustomMessagesComponent } from "./datetimepicker/localization/datetimepicker-custom-messages.component";
import { TimePickerCustomMessagesComponent } from "./timepicker/localization/timepicker-custom-messages.component";
import { ToggleButtonTabStopDirective } from "@progress/kendo-angular-common";
/**
* Utility array that contains all `Calendar` related components and directives
*/
export const KENDO_CALENDAR = [
    CalendarCustomMessagesComponent,
    CalendarComponent,
    CellTemplateDirective,
    CenturyCellTemplateDirective,
    DecadeCellTemplateDirective,
    FooterTemplateDirective,
    HeaderTemplateDirective,
    HeaderTitleTemplateDirective,
    MonthCellTemplateDirective,
    NavigationItemTemplateDirective,
    WeekNumberCellTemplateDirective,
    YearCellTemplateDirective,
    NavigationItemTemplateDirective
];
/**
* Utility array that contains all `MultiViewCalendar` related components and directives
*/
export const KENDO_MULTIVIEWCALENDAR = [
    MultiViewCalendarCustomMessagesComponent,
    MultiViewCalendarComponent,
    CellTemplateDirective,
    CenturyCellTemplateDirective,
    DecadeCellTemplateDirective,
    FooterTemplateDirective,
    HeaderTemplateDirective,
    HeaderTitleTemplateDirective,
    MonthCellTemplateDirective,
    NavigationItemTemplateDirective,
    WeekNumberCellTemplateDirective,
    YearCellTemplateDirective
];
/**
* Utility array that contains all `DateInput` related components and directives
*/
export const KENDO_DATEINPUT = [
    DateInputCustomMessagesComponent,
    DateInputComponent
];
/**
* Utility array that contains all `DatePicker` related components and directives
*/
export const KENDO_DATEPICKER = [
    DatePickerCustomMessagesComponent,
    DatePickerComponent,
    CellTemplateDirective,
    CenturyCellTemplateDirective,
    DecadeCellTemplateDirective,
    FooterTemplateDirective,
    HeaderTemplateDirective,
    HeaderTitleTemplateDirective,
    MonthCellTemplateDirective,
    NavigationItemTemplateDirective,
    WeekNumberCellTemplateDirective,
    YearCellTemplateDirective,
    NavigationItemTemplateDirective,
    ToggleButtonTabStopDirective
];
/**
* Utility array that contains all `DateRange` related components and directives
*/
export const KENDO_DATERANGE = [
    DateRangePopupCustomMessagesComponent,
    ...KENDO_DATEINPUT,
    DateRangeEndInputDirective,
    DateRangePopupTemplateDirective,
    DateRangeComponent,
    DateRangePopupComponent,
    DateRangeSelectionDirective,
    DateRangeStartInputDirective
];
/**
* Utility array that contains all `DateTimePicker` related components and directives
*/
export const KENDO_DATETIMEPICKER = [
    DateTimePickerCustomMessagesComponent,
    DateTimePickerComponent,
    CellTemplateDirective,
    CenturyCellTemplateDirective,
    DecadeCellTemplateDirective,
    FooterTemplateDirective,
    HeaderTemplateDirective,
    HeaderTitleTemplateDirective,
    MonthCellTemplateDirective,
    NavigationItemTemplateDirective,
    WeekNumberCellTemplateDirective,
    YearCellTemplateDirective,
    ToggleButtonTabStopDirective
];
/**
* Utility array that contains all `TimePicker` related components and directives
*/
export const KENDO_TIMEPICKER = [
    TimePickerCustomMessagesComponent,
    TimePickerComponent,
    ToggleButtonTabStopDirective
];
/**
* Utility array that contains all `@progress/kendo-angular-dateinputs` related components and directives
*/
export const KENDO_DATEINPUTS = [
    ...KENDO_CALENDAR,
    ...KENDO_MULTIVIEWCALENDAR,
    ...KENDO_DATEINPUT,
    ...KENDO_DATEPICKER,
    ...KENDO_DATERANGE,
    ...KENDO_DATETIMEPICKER,
    ...KENDO_TIMEPICKER
];
