// @progress/kendo-angular-dateinputs/esm2022/calendar/multiview-calendar.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_MULTIVIEWCALENDAR } from '../directives';
import { CenturyViewService } from './services/century-view.service';
import { DecadeViewService } from './services/decade-view.service';
import { MonthViewService } from './services/month-view.service';
import { NavigationService } from './services/navigation.service';
import { YearViewService } from './services/year-view.service';
import { PopupService } from '@progress/kendo-angular-popup';
import * as i0 from "@angular/core";
import * as i1 from "./localization/multiview-calendar-custom-messages.component";
import * as i2 from "./multiview-calendar.component";
import * as i3 from "./templates/cell-template.directive";
import * as i4 from "./templates/century-cell-template.directive";
import * as i5 from "./templates/decade-cell-template.directive";
import * as i6 from "./templates/footer-template.directive";
import * as i7 from "./templates/header-template.directive";
import * as i8 from "./templates/header-title-template.directive";
import * as i9 from "./templates/month-cell-template.directive";
import * as i10 from "./templates/navigation-item-template.directive";
import * as i11 from "./templates/weeknumber-cell-template.directive";
import * as i12 from "./templates/year-cell-template.directive";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * The package exports:
 * - `CellTemplateDirective`&mdash;The month cell template directive.
 * - `MonthCellTemplateDirective`&mdash;The month cell template directive.
 * - `YearCellTemplateDirective`&mdash;The year cell template directive.
 * - `DecadeCellTemplateDirective`&mdash;The decade cell template directive.
 * - `CenturyCellTemplateDirective`&mdash;The century cell template directive.
 * - `WeekNumberCellTemplateDirective`&mdash;The month week number cell template directive.
 * - `HeaderTitleTemplateDirective`&mdash;The header title template directive.
 *
 * @example
 *
 * ```ts-no-run
 * // Import the MultiViewCalendar module
 * import { MultiViewCalendarModule } from '@progress/kendo-angular-dateinputs';
 *
 * // The browser platform with a compiler
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 *
 * import { NgModule } from '@angular/core';
 *
 * // Import the app component
 * import { AppComponent } from './app.component';
 *
 * // Define the app module
 * _@NgModule({
 *     declarations: [AppComponent], // declare app component
 *     imports:      [BrowserModule, MultiViewCalendarModule], // import MultiViewCalendar module
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 *
 * // Compile and launch the module
 * platformBrowserDynamic().bootstrapModule(AppModule);
 */
export class MultiViewCalendarModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: MultiViewCalendarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: MultiViewCalendarModule, imports: [i1.MultiViewCalendarCustomMessagesComponent, i2.MultiViewCalendarComponent, i3.CellTemplateDirective, i4.CenturyCellTemplateDirective, i5.DecadeCellTemplateDirective, i6.FooterTemplateDirective, i7.HeaderTemplateDirective, i8.HeaderTitleTemplateDirective, i9.MonthCellTemplateDirective, i10.NavigationItemTemplateDirective, i11.WeekNumberCellTemplateDirective, i12.YearCellTemplateDirective], exports: [i1.MultiViewCalendarCustomMessagesComponent, i2.MultiViewCalendarComponent, i3.CellTemplateDirective, i4.CenturyCellTemplateDirective, i5.DecadeCellTemplateDirective, i6.FooterTemplateDirective, i7.HeaderTemplateDirective, i8.HeaderTitleTemplateDirective, i9.MonthCellTemplateDirective, i10.NavigationItemTemplateDirective, i11.WeekNumberCellTemplateDirective, i12.YearCellTemplateDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: MultiViewCalendarModule, providers: [
            PopupService,
            NavigationService,
            CenturyViewService,
            DecadeViewService,
            MonthViewService,
            YearViewService
        ], imports: [i2.MultiViewCalendarComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: MultiViewCalendarModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_MULTIVIEWCALENDAR],
                    imports: [...KENDO_MULTIVIEWCALENDAR],
                    providers: [
                        PopupService,
                        NavigationService,
                        CenturyViewService,
                        DecadeViewService,
                        MonthViewService,
                        YearViewService
                    ]
                }]
        }] });
