// @progress/kendo-angular-layout/esm2022/directives.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { AvatarComponent } from "./avatar/avatar.component";
import { AvatarCustomMessagesComponent } from "./avatar/l10n/custom-messages.component";
import { CardActionsComponent } from "./card/card-actions.component";
import { CardBodyComponent } from "./card/card-body.component";
import { CardFooterComponent } from "./card/card-footer.component";
import { CardHeaderComponent } from "./card/card-header.component";
import { CardComponent } from "./card/card.component";
import { CardMediaDirective } from "./card/directives/card-media.directive";
import { CardSeparatorDirective } from "./card/directives/card-separator.directive";
import { CardSubtitleDirective } from "./card/directives/card-subtitle.directive";
import { CardTitleDirective } from "./card/directives/card-title.directive";
import { DrawerContainerComponent } from "./drawer/drawer-container.component";
import { DrawerContentComponent } from "./drawer/drawer-content.component";
import { DrawerComponent } from "./drawer/drawer.component";
import { DrawerFooterTemplateDirective, DrawerHeaderTemplateDirective, DrawerItemTemplateDirective, DrawerTemplateDirective } from "./drawer/template-directives";
import { ExpansionPanelTitleDirective } from "./expansionpanel/expansionpanel-title.directive";
import { ExpansionPanelComponent } from "./expansionpanel/expansionpanel.component";
import { GridLayoutComponent } from "./layouts/grid-layout.component";
import { GridLayoutItemComponent } from "./layouts/gridlayout-item.component";
import { StackLayoutComponent } from "./layouts/stack-layout.component";
import { PanelBarContentDirective } from "./panelbar/panelbar-content.directive";
import { PanelBarItemTemplateDirective } from "./panelbar/panelbar-item-template.directive";
import { PanelBarItemTitleDirective } from "./panelbar/panelbar-item-title.directive";
import { PanelBarItemComponent } from "./panelbar/panelbar-item.component";
import { PanelBarComponent } from "./panelbar/panelbar.component";
import { SplitterPaneComponent } from "./splitter/splitter-pane.component";
import { SplitterComponent } from "./splitter/splitter.component";
import { StepperCustomMessagesComponent } from "./stepper/localization/custom-messages.component";
import { StepperComponent } from "./stepper/stepper.component";
import { StepperIndicatorTemplateDirective, StepperLabelTemplateDirective, StepperStepTemplateDirective } from "./stepper/template-directives";
import { TabContentDirective } from "./tabstrip/directives/tab-content.directive";
import { TabTitleDirective } from "./tabstrip/directives/tab-title.directive";
import { TabStripCustomMessagesComponent } from "./tabstrip/localization/custom-messages.component";
import { TabStripTabComponent } from "./tabstrip/models/tabstrip-tab.component";
import { TabStripComponent } from "./tabstrip/tabstrip.component";
import { TileLayoutItemBodyComponent } from "./tilelayout/tilelayout-item-body.component";
import { TileLayoutItemHeaderComponent } from "./tilelayout/tilelayout-item-header.component";
import { TileLayoutItemComponent } from "./tilelayout/tilelayout-item.component";
import { TileLayoutComponent } from "./tilelayout/tilelayout.component";
import { TimelineCustomMessagesComponent } from "./timeline/localization/custom-messages.component";
import { TimelineCardActionsTemplateDirective } from "./timeline/templates/timeline-card-actions.directive";
import { TimelineCardBodyTemplateDirective } from "./timeline/templates/timeline-card-body.directive";
import { TimelineCardHeaderTemplateDirective } from "./timeline/templates/timeline-card-header.directive";
import { TimelineComponent } from "./timeline/timeline.component";
/**
 * Utility array that contains all `Avatar` related components and directives.
 */
export const KENDO_AVATAR = [
    AvatarComponent,
    AvatarCustomMessagesComponent
];
/**
 * Utility array that contains all `Card` related components and directives.
 */
export const KENDO_CARD = [
    CardComponent,
    CardActionsComponent,
    CardBodyComponent,
    CardFooterComponent,
    CardHeaderComponent,
    CardMediaDirective,
    CardSeparatorDirective,
    CardSubtitleDirective,
    CardTitleDirective
];
/**
 * Utility array that contains all `Drawer` related components and directives.
 */
export const KENDO_DRAWER = [
    DrawerComponent,
    DrawerContainerComponent,
    DrawerContentComponent,
    DrawerTemplateDirective,
    DrawerFooterTemplateDirective,
    DrawerHeaderTemplateDirective,
    DrawerItemTemplateDirective
];
/**
 * Utility array that contains all `ExpansionPanel` related components and directives.
 */
export const KENDO_EXPANSIONPANEL = [
    ExpansionPanelComponent,
    ExpansionPanelTitleDirective
];
/**
 * Utility array that contains all `GridLayout` related components and directives.
 */
export const KENDO_GRIDLAYOUT = [
    GridLayoutComponent,
    GridLayoutItemComponent
];
/**
 * Utility array that contains all `PanelBar` related components and directives.
 */
export const KENDO_PANELBAR = [
    PanelBarComponent,
    PanelBarItemComponent,
    PanelBarContentDirective,
    PanelBarItemTemplateDirective,
    PanelBarItemTitleDirective
];
/**
 * Utility array that contains all `Splitter` related components and directives.
 */
export const KENDO_SPLITTER = [
    SplitterComponent,
    SplitterPaneComponent
];
/**
 * Utility array that contains all `StackLayout` related components and directives.
 */
export const KENDO_STACKLAYOUT = [
    StackLayoutComponent
];
/**
 * Utility array that contains all `Stepper` related components and directives.
 */
export const KENDO_STEPPER = [
    StepperComponent,
    StepperCustomMessagesComponent,
    StepperIndicatorTemplateDirective,
    StepperLabelTemplateDirective,
    StepperStepTemplateDirective
];
/**
 * Utility array that contains all `TabStrip` related components and directives.
 */
export const KENDO_TABSTRIP = [
    TabStripComponent,
    TabStripTabComponent,
    TabContentDirective,
    TabTitleDirective,
    TabStripCustomMessagesComponent
];
/**
 * Utility array that contains all `TileLayout` related components and directives.
 */
export const KENDO_TILELAYOUT = [
    TileLayoutComponent,
    TileLayoutItemBodyComponent,
    TileLayoutItemComponent,
    TileLayoutItemHeaderComponent
];
/**
 * Utility array that contains all `Timeline` related components and directives
 */
export const KENDO_TIMELINE = [
    TimelineComponent,
    TimelineCustomMessagesComponent,
    TimelineCardActionsTemplateDirective,
    TimelineCardBodyTemplateDirective,
    TimelineCardHeaderTemplateDirective
];
/**
 * Utility array that contains all `@progress/kendo-angular-layout` related components and directives.
 */
export const KENDO_LAYOUT = [
    ...KENDO_AVATAR,
    ...KENDO_CARD,
    ...KENDO_DRAWER,
    ...KENDO_EXPANSIONPANEL,
    ...KENDO_GRIDLAYOUT,
    ...KENDO_PANELBAR,
    ...KENDO_SPLITTER,
    ...KENDO_STACKLAYOUT,
    ...KENDO_STEPPER,
    ...KENDO_TABSTRIP,
    ...KENDO_TILELAYOUT,
    ...KENDO_TIMELINE
];
