// @progress/kendo-angular-layout/esm2022/layout.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_LAYOUT } from './directives';
import * as i0 from "@angular/core";
import * as i1 from "./avatar/avatar.component";
import * as i2 from "./avatar/l10n/custom-messages.component";
import * as i3 from "./card/card.component";
import * as i4 from "./card/card-actions.component";
import * as i5 from "./card/card-body.component";
import * as i6 from "./card/card-footer.component";
import * as i7 from "./card/card-header.component";
import * as i8 from "./card/directives/card-media.directive";
import * as i9 from "./card/directives/card-separator.directive";
import * as i10 from "./card/directives/card-subtitle.directive";
import * as i11 from "./card/directives/card-title.directive";
import * as i12 from "./drawer/drawer.component";
import * as i13 from "./drawer/drawer-container.component";
import * as i14 from "./drawer/drawer-content.component";
import * as i15 from "./drawer/template-directives/drawer-template.directive";
import * as i16 from "./drawer/template-directives/footer-template.directive";
import * as i17 from "./drawer/template-directives/header-template.directive";
import * as i18 from "./drawer/template-directives/item-template.directive";
import * as i19 from "./expansionpanel/expansionpanel.component";
import * as i20 from "./expansionpanel/expansionpanel-title.directive";
import * as i21 from "./layouts/grid-layout.component";
import * as i22 from "./layouts/gridlayout-item.component";
import * as i23 from "./panelbar/panelbar.component";
import * as i24 from "./panelbar/panelbar-item.component";
import * as i25 from "./panelbar/panelbar-content.directive";
import * as i26 from "./panelbar/panelbar-item-template.directive";
import * as i27 from "./panelbar/panelbar-item-title.directive";
import * as i28 from "./splitter/splitter.component";
import * as i29 from "./splitter/splitter-pane.component";
import * as i30 from "./layouts/stack-layout.component";
import * as i31 from "./stepper/stepper.component";
import * as i32 from "./stepper/localization/custom-messages.component";
import * as i33 from "./stepper/template-directives/indicator-template.directive";
import * as i34 from "./stepper/template-directives/label-template.directive";
import * as i35 from "./stepper/template-directives/step-template.directive";
import * as i36 from "./tabstrip/tabstrip.component";
import * as i37 from "./tabstrip/models/tabstrip-tab.component";
import * as i38 from "./tabstrip/directives/tab-content.directive";
import * as i39 from "./tabstrip/directives/tab-title.directive";
import * as i40 from "./tabstrip/localization/custom-messages.component";
import * as i41 from "./tilelayout/tilelayout.component";
import * as i42 from "./tilelayout/tilelayout-item-body.component";
import * as i43 from "./tilelayout/tilelayout-item.component";
import * as i44 from "./tilelayout/tilelayout-item-header.component";
import * as i45 from "./timeline/timeline.component";
import * as i46 from "./timeline/localization/custom-messages.component";
import * as i47 from "./timeline/templates/timeline-card-actions.directive";
import * as i48 from "./timeline/templates/timeline-card-body.directive";
import * as i49 from "./timeline/templates/timeline-card-header.directive";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the Layout components.
 *
 * @example
 *
 * ```ts-no-run
 * // Import the Layout module
 * import { LayoutModule } from '@progress/kendo-angular-layout';
 *
 * // The browser platform with a compiler
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 *
 * import { NgModule } from '@angular/core';
 *
 * // Import the app component
 * import { AppComponent } from './app.component';
 *
 * // Define the app module
 * _@NgModule({
 *     declarations: [AppComponent], // declare app component
 *     imports:      [BrowserModule, LayoutModule], // import Layout module
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 *
 * // Compile and launch the module
 * platformBrowserDynamic().bootstrapModule(AppModule);
 *
 * ```
 */
export class LayoutModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: LayoutModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: LayoutModule, imports: [i1.AvatarComponent, i2.AvatarCustomMessagesComponent, i3.CardComponent, i4.CardActionsComponent, i5.CardBodyComponent, i6.CardFooterComponent, i7.CardHeaderComponent, i8.CardMediaDirective, i9.CardSeparatorDirective, i10.CardSubtitleDirective, i11.CardTitleDirective, i12.DrawerComponent, i13.DrawerContainerComponent, i14.DrawerContentComponent, i15.DrawerTemplateDirective, i16.DrawerFooterTemplateDirective, i17.DrawerHeaderTemplateDirective, i18.DrawerItemTemplateDirective, i19.ExpansionPanelComponent, i20.ExpansionPanelTitleDirective, i21.GridLayoutComponent, i22.GridLayoutItemComponent, i23.PanelBarComponent, i24.PanelBarItemComponent, i25.PanelBarContentDirective, i26.PanelBarItemTemplateDirective, i27.PanelBarItemTitleDirective, i28.SplitterComponent, i29.SplitterPaneComponent, i30.StackLayoutComponent, i31.StepperComponent, i32.StepperCustomMessagesComponent, i33.StepperIndicatorTemplateDirective, i34.StepperLabelTemplateDirective, i35.StepperStepTemplateDirective, i36.TabStripComponent, i37.TabStripTabComponent, i38.TabContentDirective, i39.TabTitleDirective, i40.TabStripCustomMessagesComponent, i41.TileLayoutComponent, i42.TileLayoutItemBodyComponent, i43.TileLayoutItemComponent, i44.TileLayoutItemHeaderComponent, i45.TimelineComponent, i46.TimelineCustomMessagesComponent, i47.TimelineCardActionsTemplateDirective, i48.TimelineCardBodyTemplateDirective, i49.TimelineCardHeaderTemplateDirective], exports: [i1.AvatarComponent, i2.AvatarCustomMessagesComponent, i3.CardComponent, i4.CardActionsComponent, i5.CardBodyComponent, i6.CardFooterComponent, i7.CardHeaderComponent, i8.CardMediaDirective, i9.CardSeparatorDirective, i10.CardSubtitleDirective, i11.CardTitleDirective, i12.DrawerComponent, i13.DrawerContainerComponent, i14.DrawerContentComponent, i15.DrawerTemplateDirective, i16.DrawerFooterTemplateDirective, i17.DrawerHeaderTemplateDirective, i18.DrawerItemTemplateDirective, i19.ExpansionPanelComponent, i20.ExpansionPanelTitleDirective, i21.GridLayoutComponent, i22.GridLayoutItemComponent, i23.PanelBarComponent, i24.PanelBarItemComponent, i25.PanelBarContentDirective, i26.PanelBarItemTemplateDirective, i27.PanelBarItemTitleDirective, i28.SplitterComponent, i29.SplitterPaneComponent, i30.StackLayoutComponent, i31.StepperComponent, i32.StepperCustomMessagesComponent, i33.StepperIndicatorTemplateDirective, i34.StepperLabelTemplateDirective, i35.StepperStepTemplateDirective, i36.TabStripComponent, i37.TabStripTabComponent, i38.TabContentDirective, i39.TabTitleDirective, i40.TabStripCustomMessagesComponent, i41.TileLayoutComponent, i42.TileLayoutItemBodyComponent, i43.TileLayoutItemComponent, i44.TileLayoutItemHeaderComponent, i45.TimelineComponent, i46.TimelineCustomMessagesComponent, i47.TimelineCardActionsTemplateDirective, i48.TimelineCardBodyTemplateDirective, i49.TimelineCardHeaderTemplateDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: LayoutModule, imports: [i1.AvatarComponent, i12.DrawerComponent, i19.ExpansionPanelComponent, i23.PanelBarComponent, i24.PanelBarItemComponent, i28.SplitterComponent, i31.StepperComponent, i36.TabStripComponent, i41.TileLayoutComponent, i45.TimelineComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: LayoutModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_LAYOUT],
                    imports: [...KENDO_LAYOUT]
                }]
        }] });
