// @progress/kendo-angular-inputs/esm2022/textarea/textarea-suffix.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Component, HostBinding, Input } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Specifies the adornments in the suffix container ([see example]({% slug textarea_adornments %})).
 * ```ts-no-run
 * _@Component({
 * selector: 'my-app',
 * template: `
 *  <kendo-textarea>
 *    <kendo-textarea-suffix>
 *        <button kendoButton look="clear" icon="image"></button>
 *    </kendo-textarea-suffix>
 *  </kendo-textarea>
 * `
 * })
 * class AppComponent {}
 * ```
 */
export class TextAreaSuffixComponent {
    /**
     * @hidden
     */
    flow = 'vertical';
    /**
     * @hidden
     */
    orientation = 'horizontal';
    hostClass = true;
    get verticalOrientation() {
        return this.orientation === 'vertical';
    }
    get horizontalOrientation() {
        return this.orientation === 'horizontal';
    }
    get alignItems() {
        return this.flow === this.orientation;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TextAreaSuffixComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: TextAreaSuffixComponent, isStandalone: true, selector: "kendo-textarea-suffix", inputs: { flow: "flow", orientation: "orientation" }, host: { properties: { "class.k-input-suffix": "this.hostClass", "class.k-input-suffix-vertical": "this.verticalOrientation", "class.k-input-suffix-horizontal": "this.horizontalOrientation", "class.!k-align-items-start": "this.alignItems" } }, exportAs: ["kendoTextAreaSuffix"], ngImport: i0, template: `<ng-content></ng-content>`, isInline: true });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TextAreaSuffixComponent, decorators: [{
            type: Component,
            args: [{
                    exportAs: 'kendoTextAreaSuffix',
                    selector: 'kendo-textarea-suffix',
                    template: `<ng-content></ng-content>`,
                    standalone: true
                }]
        }], propDecorators: { flow: [{
                type: Input
            }], orientation: [{
                type: Input
            }], hostClass: [{
                type: HostBinding,
                args: ['class.k-input-suffix']
            }], verticalOrientation: [{
                type: HostBinding,
                args: ['class.k-input-suffix-vertical']
            }], horizontalOrientation: [{
                type: HostBinding,
                args: ['class.k-input-suffix-horizontal']
            }], alignItems: [{
                type: HostBinding,
                args: ['class.!k-align-items-start']
            }] } });
